tsinfer/inference.py

```diff
@@ -2179,3 +2298,93 @@ def minimise(ts):
         filter_individuals=False,
         filter_populations=False,
     )
+
+
+def solve_num_mismatches(ts, k):
+    """
+    Return the low-level LS parameters corresponding to accepting
+    k mismatches in favour of a single recombination.
+    """
+    m = ts.num_sites
+    n = ts.num_nodes  # We can match against any node in tsinfer
+    if k == 0:
+        # Pathological things happen when k=0
+        r = 1e-3
+        mu = 1e-20
+    else:
+        mu = 1e-6
+        denom = (1 - mu) ** k + (n - 1) * mu**k
+        r = n * mu**k / denom
+        # print("n = ", n, "k = ",k)
+        # print(mu, r)
+        assert mu < 0.5
+        assert r < 0.5
+
+    ls_recomb = np.full(m - 1, r)
+    ls_mismatch = np.full(m, mu)
+    return ls_recomb, ls_mismatch
+
+
+class SequentialExtender:
+    def __init__(self, sample_data, ancestors_ts=None, time_units=None):
+        time_units = tskit.TIME_UNITS_UNCALIBRATED if time_units is None else time_units
+        self.sample_data = sample_data
+        if ancestors_ts is None:
+            tables = tskit.TableCollection(sample_data.sequence_length)
+            tables.time_units = time_units
+            for site in sample_data.sites():
+                tables.sites.add_row(site.position, site.ancestral_state)
+            tables.nodes.metadata_schema = tskit.MetadataSchema.permissive_json()
+            # TODO should probably make the ultimate ancestor time something less
+            # plausible or at least configurable.
+            for t in [1, 0]:
+                tables.nodes.add_row(time=t)
+            tables.edges.add_row(0, sample_data.sequence_length, 0, 1)
+            self.ancestors_ts = tables.tree_sequence()
+
+            # Seed the known_haplotypes with the all-zeros haplotypes so that
+            # we don't create a spurious node for it in the first generation.
+            root_haplotype = next(sample_data.haplotypes())[1].copy()
+            root_haplotype[:] = 0
+            self.haplotypes = {root_haplotype.tobytes()}
+        else:
+            self.ancestors_ts = ancestors_ts
+            if time_units is not None and self.ancestors_ts.time_units != time_units:
+                raise ValueError(
+                    f"Mismatched time_units: {time_units} != ancestors_ts.time_units",
+                )
+            # Add in the existing haplotypes. Note - this will probably
```

Comment:
At scale, we probably don't want to match against the sample haplotypes in the next round of extend, if there is a simple "exemplar" we can match against instead. But there's no reason why we can't have these in the actual TS. Is there an argument here for a flag "NODE_HAS_OLDER_PROXY", which would simply cause the matching algorithm to skip matching against those nodes, since we are guaranteed that we can match against an older haplotype anyway?